engine/src/main/java/org/hibernate/validator/constraints/UUID.java

```diff
@@ -0,0 +1,144 @@
+ */
+package org.hibernate.validator.constraints;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.hibernate.validator.constraints.UUID.List;
+
+import jakarta.validation.Constraint;
+import jakarta.validation.Payload;
+
+import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
+import static java.lang.annotation.ElementType.CONSTRUCTOR;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.ElementType.TYPE_USE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+
+/**
+ * Checks that the annotated character sequence is a valid
+ * <a href="https://en.wikipedia.org/wiki/Universally_unique_identifier">UUID</a>.
+ * <p>
+ * Validation characteristics:
+ * <ul>
+ * <li>Consists only of numbers, hex characters and dashes</li>
+ * <li>Has exact length of 36 characters</li>
+ * <li>Nil UUID (default: allowed)</li>
+ * <li>UUID version minimum and maximum (default: minimum version of 1 and maximum version of 5)</li>
+ * <li>UUID variant minimum and maximum (default: minimum variant of 0 and maximum variant of 2)</li>
+ * <li>Letter case (default: lower case)</li>
+ * </ul>
+ * <p>
+ * The supported type is {@code CharSequence}. {@code null} is considered valid.
+ *
+ * @author Daniel Heid
+ * @since 7.0.3
+ */
+@Documented
+@Constraint(validatedBy = {})
+@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+@Retention(RUNTIME)
+@Repeatable(List.class)
+public @interface UUID {
+
+	String message() default "{org.hibernate.validator.constraints.UUID.message}";
+
+	Class<?>[] groups() default {};
+
+	Class<? extends Payload>[] payload() default {};
+
+	/**
+	 * @return allow empty strings.
+	 * Per default does not allow empty strings
+	 */
+	boolean allowEmpty() default false;
+
+	/**
+	 * @return {@code true} if nil UUIDs {@code 00000000-0000-0000-0000-000000000000} are valid
+	 * Per default nil UUIDs are valid
+	 */
+	boolean allowNil() default true;
+
+	/**
+	 * Must not be lower than version 1
+	 *
+	 * @return the minimum UUID version number
+	 * Per default version 1 should be minimum
+	 */
+	int minimumVersion() default 1;
+
+	/**
+	 * Must not be lower than version 1
+	 *
+	 * @return the maximum UUID version number
+	 * Per default version 5 should be maximum
+	 */
+	int maximumVersion() default 5;
+
+	/**
+	 * Must not be lower than variant 0
+	 *
+	 * @return the minimum UUID variant number
+	 * Per default variant 0 should be minimum
+	 */
+	int minimumVariant() default 0;
+
+	/**
+	 * Must not be lower than variant 0
+	 *
+	 * @return the maximum UUID variant number
+	 * Per default variant 2 should be maximum
+	 */
+	int maximumVariant() default 2;
```

Comment:
Considering there are very few valid versions (1-5) and variants (0-2), wouldn't it be better to have `int[] version() default {1, 2, 3, 4, 5};` and `int[] variant() default {0, 1, 2}`?

That way a user can allow version 1 and 5 but not 2, 3, 4, for example.

Reply:
Good idea! I will change that

Reply:
@yrodiere Done! :smile: